File: arena-junit/src/test/java/arena/junit/dep/oracledb/OracleDependencyBuilderSerializationTest.java
package arena.junit.dep.oracledb;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import arena.junit.match.ArenaRunnableDependency;
import arena.junit.support.ArenaJson;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;
import org.junit.jupiter.api.Test;

final class OracleDependencyBuilderSerializationTest {

  static final class StubDependency implements ArenaRunnableDependency {
    @Override
    public ObjectNode forFfi() {
      return ArenaJson.object().put("identifier", "child");
    }
  }

  @Test
  void build_minimalName_serializesTypeAndIdentifier() {
    ObjectNode config = new OracleDependencyBuilder("ora").build().forFfi();
    assertEquals("oracle", config.path("type").asText());
    assertTrue(config.path("identifier").asText().startsWith("arena-oracle-ora-"));
    assertFalse(config.has("children"));
  }

  @Test
  void build_minimalName_omitsAdminPasswordKey() {
    ObjectNode config = new OracleDependencyBuilder("ora").build().forFfi();
    assertFalse(config.has("admin_password"));
  }

  @Test
  void withOverrides_allFields_serializesConfiguredFields() {
    ObjectNode config =
        new OracleDependencyBuilder("ora")
            .withImage("21-slim")
            .withImageName("oracle-free")
            .withPort(1521)
            .withDatabaseName("arena")
            .withDatabaseUsername("arena_user")
            .withDatabasePassword("secret")
            .withAdminPassword("secret-admin")
            .withContainerName("ora-box")
            .build()
            .forFfi();
    assertEquals("21-slim", config.path("image").asText());
    assertEquals("oracle-free", config.path("image_name").asText());
    assertEquals(1521, config.path("port").asInt());
    assertEquals("arena", config.path("database_name").asText());
    assertEquals("arena_user", config.path("database_username").asText());
    assertEquals("secret", config.path("database_password").asText());
    assertEquals("secret-admin", config.path("admin_password").asText());
    assertEquals("ora-box", config.path("container_name").asText());
  }

  @Test
  void withStartupSqlScripts_multipleScripts_serializesInOrder() {
    ObjectNode config =
        new OracleDependencyBuilder("ora")
            .withStartupSqlScripts(List.of("seed.sql", "grants.sql"))
            .build()
            .forFfi();
    ArrayNode scripts = (ArrayNode) config.path("startup_sql_scripts");
    assertEquals(2, scripts.size());
    assertEquals("seed.sql", scripts.get(0).asText());
    assertEquals("grants.sql", scripts.get(1).asText());
  }

  @Test
  void addChildDependency_stubDependency_serializesChildrenViaChildrenFfi() {
    ObjectNode config =
        new OracleDependencyBuilder("ora")
            .addChildDependency(new StubDependency())
            .build()
            .forFfi();
    assertEquals(1, ((ArrayNode) config.path("children")).size());
  }

  @Test
  void identifier_afterBuild_returnsConfiguredIdentifier() {
    OracleDependency dep = new OracleDependencyBuilder("ora").build();
    assertTrue(dep.identifier().startsWith("arena-oracle-ora-"));
  }
}

File: arena-xunit/tests/Unit/OracleDependencyBuilderSerializationTest.cs
using System;
using System.Linq;
using ArenaDotnet.Xunit.Dep;
using ArenaDotnet.Xunit.Support;
using Newtonsoft.Json.Linq;
using Xunit;

namespace ArenaDotnet.Xunit.UnitTest;

public class OracleDependencyBuilderSerializationTest
{
    [Fact]
    public void Build_DefaultPort_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("oracle", obj["type"]);
        Assert.Equal(1521, obj["port"]);
        Assert.NotNull(obj["identifier"]);
    }

    [Fact]
    public void Build_CustomPort_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").WithPort(1522).Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal(1522, obj["port"]);
    }

    [Fact]
    public void Build_Identifier_MatchesPattern()
    {
        var dep = new OracleDependencyBuilder("test").Build();
        Assert.StartsWith("arena-oracle-", dep.Identifier);
    }

    [Fact]
    public void Build_WithImageName_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").WithImageName("oracle-free").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("oracle-free", obj["image_name"]);
    }

    [Fact]
    public void Build_WithImage_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").WithImage("21-slim").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("21-slim", obj["image"]);
    }

    [Fact]
    public void Build_WithContainerName_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").WithContainerName("my-oracle").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("my-oracle", obj["container_name"]);
    }

    [Fact]
    public void Build_WithAdminPassword_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test").WithAdminPassword("secret-admin").Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("secret-admin", obj["admin_password"]);
    }

    [Fact]
    public void Build_WithDatabaseFieldsAndStartupScripts_SerializesCorrectJson()
    {
        var dep = new OracleDependencyBuilder("test")
            .WithDatabaseName("arena")
            .WithDatabaseUsername("arena_user")
            .WithDatabasePassword("secret")
            .WithStartupSqlScripts(new[] { "seed.sql", "grants.sql" })
            .Build();
        var json = dep.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("arena", obj["database_name"]);
        Assert.Equal("arena_user", obj["database_username"]);
        Assert.Equal("secret", obj["database_password"]);
        var scripts = Assert.IsType<JArray>(obj["startup_sql_scripts"]);
        Assert.Equal(2, scripts.Count);
        Assert.Equal("seed.sql", scripts[0]);
        Assert.Equal("grants.sql", scripts[1]);
    }
}
